### DISCUSSION TITLE
Preserve HTTP params during navigation

### DISCUSSION THREAD

**Discussion by vporton** [remix-run/react-router#12058] [Proposals]
I am creating a site. Every page of the site goes with `canisterId=XXX` GET param and optional `backend=YYY` param.

So, I want `ReactBrowser`'s `navigate()` method as well as Link tags to preserve these GET params.

For example, `<Link to="/bookmark">` should lead to `/bookmark?canisterId=XXX` or `/bookmark?canisterId=XXX&backend=YYY` rather that to `/bookmark`.

Is it implemented? How to do it?

I suggest to add to `(Browser)Router` attribute `persistentParams` like `persistentParams={{canisterId: true}}`. Alternatively it could be done as `persistentParamsCallback={f}` where `f` receives a param and returns a boolean whether it should persist.

**Comment by kiliman** [comment]
You'll probably have better luck creating your own wrapper `<Link>` component.

Also, what happens if you have a URL with a search param? Do you append the existing ones as well? What about if you have duplicates? Does one overwrite the other, or do you append both?

`<Link to="/bookmark?key=value">`

**Comment by vporton** [comment]
Do you append the existing ones as well? - If I have `<Link to="/aaa?search=bbb">`, it would become `<Link to="/aaa?canisterId=XXX&search=bbb">`.

Not sure how to handle duplicates. For my particular application it does not matter how duplicates are handled.
